### PR TITLE
[RSDK-7680] [APP-4933] Allow non-internet wifi to count as "connected", add roaming mode

### PR DESCRIPTION
**Pull request by Otterverse**
This modifies the behavior to consider a connected wifi network as good enough, and thus prevent a hotspot from being activated, even if that wifi doesn't have internet connectivity. The previous behavior is available by enabling roaming mode in the config.

This PR also refactors a lot of the code for readability and simplification to support both modes, and merges the previously-separate "portal" structs into the main networkmanger struct to avoid data duplication.

As such, it will almost certainly be easier to review "from scratch" rather than by the file diff views.

For testing, the follow config chunk should get you started:

```
    "agent-provisioning": {
      "attributes": {
        "offline_timeout": "2m",
        "user_timeout": "5m",
        "fallback_timeout": "10m",
        "roaming_mode": false
      },
      "pin_url": "https://storage.googleapis.com/packages.viam.com/temp/viam-agent-provisioning-roamingpreview3-aarch64"
    },
```
The attributes are all the newly added options, with their default values. You can adjust them (especially roaming_mode) to make sure things still work as expected.

UPDATE: Updated pin_url to version 2 (6/18/2024), so update to new version if you already started testing with the old one.

UPDATE2: Updated pin_url to version 3 (6/24/2024). Includes minor fix around cold-start timeouts.

### CODE REVIEW

**Comment by Otterverse** [comment]
Mostly needing/wanting some extra testing from folks on this. It's working in my local use cases on a Pi4, in both the new single mode and the old roaming mode at this point. Tried breaking it in several ways with manual manipulation of NetworkManager configurations too. I ran out of scenarios to test on my own in this limited setting, so please try to break it yourselves!

File: networkmanager/networkmanager.go [hunk 1]
```diff
@@ -930,49 +554,137 @@ func (w *NMWrapper) startStateMonitors(ctx context.Context) {
...
+				userInputReceived = true
+				newSSID = cfg.SSID
+			}
+
 			// wait 3 seconds so responses can be sent to/seen by user
 			if !provisioning.HealthySleep(ctx, time.Second*3) {
 				return nil
 			}
-		case <-time.After(loopDelay):
+			if newNetwork {
```
Comment (ale7714) [inline networkmanager/networkmanager.go:610]: i think i understand the code but want to double check. This if branch only plans to networks that we're adding for the first time. Which means that on first boot we will be able to connect to the internet regardless of roaming mode.

**Review by ale7714 — COMMENTED** [review]
I did some initial testing and it looks good but i would like to do more testing

File: networkmanager/networkmanager.go (hunk 1 above)
Comment (Otterverse) [reply networkmanager/networkmanager.go:610]: I think that's correct, but I'll restate for clarity.

This branch only activates on a NEW (to network manager itself!) wifi being added, and checks/requires that it has internet connectivity. If not, it restarts provisioning and leaves a warning for the user. This happens regardless of mode. (Though roaming mode ALWAYS requires internet anyway, so behavior is more or less the same there.)

**Comment by Otterverse** [comment]
Updated pin_url to version 2 (6/18/2024), so update to new version if you already started testing with the old one.

**Comment by Otterverse** [comment]
@ale7714 Addressed your logging concerns I believe, and implemented a "find highest priority" scheme for "primary network" rather than just 999 as a magic number. Also improved some locking issues I think.

**Review by ale7714 — APPROVED** [review]
I have tried both single mode and roaming mode and it works on my pi. For this change, i think docs is going to be pretty important. We have a lot more knobs now that will be pretty useful.

**Review by ale7714 — APPROVED** [review]
👍

**Pull request merged**